### PR TITLE
feat(optimizer): support layer gradient norm

### PR DESCRIPTION
**Pull request by JiaoPL**
## Motivation
1. Support layer gradient norm and use it instead of total gradient norm.
2. Add layer gradient norm to Tensorboard.

## Modification

`internlm/core/engine.py`
`internlm/solver/optimizer/hybrid_zero_optim.py`
`internlm/solver/optimizer/utils.py`
`internlm/train/training_internlm.py`
`train.py`

- [x] Align gradient norm precision.
- [x] Test on the 7B case and the MoE case.

### CODE REVIEW

**Comment by JiaoPL** [comment]
tensorboard example:

image

File: internlm/solver/optimizer/utils.py [hunk 1]
```diff
@@ -507,18 +523,24 @@ def __init__(self, model: Union[nn.Module, nn.ModuleList]) -> None:
...
+                            layer_name = f"{block.__class__.__name__}.{idx}"
+                            global_layer_norms[layer_name] = 0.0
+                            parameter.__setattr__("layer_name", layer_name)
                 else:
                     # record the block that a parameter belongs to
                     # self._block_to_param[name] = list(children.parameters())
                     self._block_to_param[children] = list(children.parameters())
+                    for parameter in self._block_to_param[children]:
+                        layer_name = f"{children.__class__.__name__}"
+                        parameter.__setattr__("layer_name", name)
```
Comment (JiaoPL) [inline internlm/solver/optimizer/utils.py:543]: 感觉得换个地方，如果没有使用ParamBcasrSyncHandler，就G了
